Decrease columns and refactor get pods layout

andronat opened the pull request:
Refs to #7843

_Requirements_:
- NAME
  - Never abbreviated because, as Brian put it, get's primary purpose should be to provide clues about which pods need to be described. Really hard to describe a pod that you see in get if you don't have the full name. Also if we abbreviate it, it breaks all scripting where you would try to pipe the name based on some grep.
- READY
  - The number of ready containers of the total number of containers in the format M / N, where "ready" is determined by the container running and, if the container has a readiness probe, that the probe has ready status.
- REASON
  - One-word CamelCase reason:
    - If no host is assigned, Unscheduled
    - If the pod terminated with success, Success
    - If M < N,
      - and all containers terminated with success, Success
      - and >0 containers are in waiting state, pick one of the non-empty Reasons (TBD which is most relevant)
      - if no reason on any container, ExitCodeN/SignalN (need to flesh this out more)
    - If M = N, Reason should be blank
- RESTARTS
  - RestartCount of all containers summed together.
- AGE
  - Time pod was created. I would add this because (a) I think it's important when you're trying to find out which pod was just created or if you're in a created/vacated pod loop and (b) I think we can do the column in max 4 characters:
    - 1s, 59s, 1m, 59m, 1h, 23h, 1d, 364d, 1y, etc.

Can someone comment if this is on the right track?

I also have some questions:
1. Why pod.Status.ContainerStatuses has len()==0 when pod.Status.Phase == pending? Is this the correct behavior?
2. Do we want somehow to include also LastTerminationState? And if we want it, then how (level of importance, what overwrites what)?
3. What about tables that @fabianofranz proposed?
4. (a little bit irrelevant) maybe Namespace become a column?

/cc @bgrant0607 @ghodss

andronat committed:
Refactor get pods layout

bgrant0607 commented:
Sample output?

Namespace doesn't need to be column, because only objects of one namespace will be shown, at least the vast majority of the time.

bgrant0607 commented:
pod.Status.ContainerStatuses gets populated by Kubelet once the containers have been created. If the pod is pending, the containers don't exist yet.

andronat commented:
e.g.

```
$kubectl get pods
NAME             READY     REASON    RESTARTS   AGE
my-nginx-00jsd   0/1       Pending   0          ~1m
my-nginx-7kqqp   0/1       Pending   0          ~1m
my-nginx-fb2s6   0/1       Pending   0          ~1m
```

bgrant0607 commented:
Nice.

While I do think that anyone using kubectl for durable/reusable scripts needs to use templates and specify the API version, there's value in retaining the ability to do simple one-line scripts, such as to sort, grep, cut, etc., esp. until we implement that functionality. For instance, I could imagine someone doing something like:
```
kubectl describe pods $(kubectl get pods -lapp=foo | tail +1 | sort -n -r -k 4 | cut -f 1)
```

I also personally don't like the table with the row/column borders aesthetically.

bgrant0607 commented:
As for LastTerminatedState, we should add that to [describeContainers](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/describe.go#L339), by pulling out the state switch statement into another function, which can then be invoked twice.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -353,103 +352,63 @@ func podHostString(host, ip string) string {
-		return "Terminated", translateTimestamp(state.Terminated.StartedAt), getTermMsg(state.Terminated), nil
-	}
-	return "", "", "", fmt.Errorf("unknown container state %#v", *state)
+	shortHumanDuration := func(d time.Duration) string {
```
Just pull this out into an actual function. I'm sure we'll need it elsewhere.

ghodss replied:
That would also allow you to document its behavior.

I think the output could be a little simplified:

0s, 1s, 59s, 1m, 59m, 1h, 23h, 1d, 364d, 1y.

I don't think you need ~ or <. I also think you want to update at 23h, not 48h, to keep consistency.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -353,103 +352,63 @@ func podHostString(host, ip string) string {
+	readyContainers := 0
+	reason := string(pod.Status.Phase)
+
+	for _, container := range pod.Status.ContainerStatuses {
```
If reason is Succeeded, you shouldn't loop over the statuses.

andronat replied:
I think I still need to loop over statuses or else how can I take the restart counts?

bgrant0607 replied:
Good point.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -353,103 +352,63 @@ func podHostString(host, ip string) string {
+		if container.Ready {
+			readyContainers++
+		} else if container.State.Terminated != nil && container.State.Terminated.Reason != "" {
+			reason = container.State.Terminated.Reason
```
(Here and elsewhere) You probably should check whether Reason is non-empty.

bgrant0607 replied:
Oops, nevermind.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -353,103 +352,63 @@ func podHostString(host, ip string) string {
+			readyContainers++
+		} else if container.State.Terminated != nil && container.State.Terminated.Reason != "" {
+			reason = container.State.Terminated.Reason
+		} else if container.State.Waiting != nil {
```
Check whether the reason is non-empty.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -353,103 +352,63 @@ func podHostString(host, ip string) string {
+			reason = container.State.Terminated.Reason
+		} else if container.State.Waiting != nil {
+			reason = container.State.Waiting.Reason
 		}
```
There are also cases like:
https://github.com/GoogleCloudPlatform/kubernetes/blob/233ce46b83aeb30b07b52ddaddf70f7ef4148ec1/pkg/kubelet/kubelet.go#L1523

We could add a Reason to the API for this, in addition to the message.

andronat replied:
I see the point, so I will add the pod message if there is any until we add the camelCase Reason to the `pod.Status`.

By the way I also noticed [this line](https://github.com/GoogleCloudPlatform/kubernetes/blob/d0daabb34bb87f4184d25b86ab39d04da3182161/plugin/pkg/scheduler/scheduler.go#L125). The messages from scheduler are not stored in `pod.Status.Message` is this a bug? Because it is the opposite from your example with the disk space error.

bgrant0607 replied:
We don't currently have a way for the scheduler to update pod status, unfortunately. You could file an issue for that.

bgrant0607 commented:
Any comments @ghodss ?

bgrant0607 commented:
@andronat I would like to get this in by Friday.

andronat commented:
I am on it

andronat commented:
Please take a look. 

I am not sure on what I must do with `interpretContainerStatus()`. It is not currently used but if you think that we might need it in the future maybe we can store it as a comment??

ghodss commented on pkg/kubectl/resource_printer.go:
```diff
@@ -350,106 +349,61 @@ func podHostString(host, ip string) string {
+		if container.Ready {
+			readyContainers++
+		} else if container.State.Terminated != nil && container.State.Terminated.Reason != "" {
+			reason = container.State.Terminated.Reason
```
As of right now this seems to have the behavior that "reason" will simply be the reason of the last non-ready container, instead of determined by some heuristic. That's fine, but maybe document that in a comment to make it clear.

bgrant0607 replied:
I'd pick the first one instead, since users will probably list containers in order of importance.

And check Terminated before Ready, as per the comment above.

bgrant0607 replied:
More specifically, I'd choose the first Waiting reason, then the first Terminated reason.

bgrant0607 replied:
If it's terminated with no reason, that's where you could use logic similar to interpretContainerStatus, to generate ExitCodeN / SignalN.

ghodss commented:
The code LGTM. Very excited about this. We can tweak the exact way "Reason" gets generated in subsequent PR's.

On `interpretContainerStatus()`, if we don't use it anywhere anymore I'd say delete it - we can always add it back from source control later.

Also, shippable is failing but when I go to https://app.shippable.com/builds/5570e53c0d553e0c002c4946 to see the error message I get "The build you're looking for wasn't found." @bgrant0607 any idea what's up?

bgrant0607 commented:
@ghodss That message from Shippable appears to be temporary rather than permanent. When I get that, I try again and it usually works.

As for the error:
./pkg/kubectl/resource_printer_test.go fails gofmt

Install the precommit hooks, which include verify-gofmt.sh:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/development.md#hooks

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -350,106 +349,61 @@ func podHostString(host, ip string) string {
+	reason := string(pod.Status.Phase)
+
+	// TODO: replace pod.Status.Message with pod.Status.Reason (to be added soon)
+	if pod.Status.Message != "" {
```
The problem with this is that Message is not one word.

You could search the message for key words and generate reasons. Kubelet should really generate reasons. @dchen1107 

Current messages:
```
Message: "Pod cannot be started due to lack of disk space."})
```
If Message contains "disk" then reason should be OutOfDisk.
```
Message: "Pod cannot be started due to host port conflict"})
```
If Message contains "port" then reason should be HostPortConflict.
```
Message: "Pod cannot be started due to node selector mismatch"})
```
If Message contains "selector" then reason should be NodeSelectorMismatch.
```
Message: "Pod cannot be started due to exceeded capacity"})
```
If Message contains "capacity" then reason should be CapacityExceeded.
```
Message: "Pod was active on the node longer than specified deadline"}, nil
```
If message contains "deadline" then reason should be DeadlineExceeded.
```
Message: fmt.Sprintf("Query container info failed with error (%v)", err),
```
No useful reason -- choose something else.

bgrant0607 replied:
If you want to defer this to a subsequent PR, just don't initialize reason to message right now.

andronat replied:
I think I will just skip message for now as I am already working on putting Reason into Pod.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -350,106 +349,61 @@ func podHostString(host, ip string) string {
+
+	for _, container := range pod.Status.ContainerStatuses {
+		restarts += container.RestartCount
+		if container.Ready {
```
Ready might be true even if the container is terminated. @dchen1107 @mikedanese

bgrant0607 commented:
Re. interpretContainerStatus: other than the ExitCodeN / SignalN logic, explaining the LastTerminationState could be moved to describe pod.

andronat commented:
* removed pod message (to be done on a next PR)
* pick `Reason` from the first container
* reordered `Terminated` and `Waiting` reasons
* added `ExitCodeN / SignalN `
* checking if `State.Running != nil` to report ready

Fixed tests and rebased. For some reason hooks/pre-commit reports that everything is ok to me :confused:

bgrant0607 commented:
Please gofmt pkg/kubectl/resource_printer_test.go

bgrant0607 commented:
What Go version are you using? gofmt changes with each version.

andronat commented:
```
$ hooks/pre-commit
Checking for files that need gofmt... OK
Checking for files that need boilerplate... OK
Checking for API descriptions... OK
Checking for docs that need updating... OK
Checking for conversions that need updating... OK
Checking for deep-copies that need updating... OK
```

```
$ gofmt pkg/kubectl/resource_printer_test.go > fmttest
$ diff pkg/kubectl/resource_printer_test.go fmttest
$ # No diff :/
```

```
$ go version
go version go1.4.2 linux/amd64
```

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -350,106 +349,60 @@ func podHostString(host, ip string) string {
+		} else if container.State.Terminated != nil && container.State.Terminated.Reason != "" {
+			reason = container.State.Terminated.Reason
+		} else if container.State.Terminated != nil && container.State.Terminated.Reason == "" {
+			reason = fmt.Sprintf("ExitCode:%d/Signal:%d", container.State.Terminated.ExitCode, container.State.Terminated.Signal)
```
Please just print Signal:%d if Signal != 0, and ExitCode:%d otherwise.

andronat commented:
Like this?

andronat commented:
Changed `Signal` and `ExitCode`

bgrant0607 commented:
LGTM! Thanks!

We'll see whether we still get gofmt complaints.

bgrant0607 commented:
gofmt still fails. Could you maybe have an old gofmt in your path somewhere? I've been bit by that in the past.

bgrant0607 commented:
Try `gofmt -s`

bgrant0607 commented:
Please also squash the commits.

andronat commented:
OK `-s` worked. The hooks did't work because I had already committed the files and the `pre-commit` script checks only what ever is in cached. I had an older Go version in the beginning when I wrote those tests so that confused the situation....

bgrant0607 commented:
Thanks.
Please squash commits.
```
git rebase -i HEAD~3
```
Change the 2nd two commits from "pick" to "fixup".
https://help.github.com/articles/about-git-rebase/#commands-available-while-rebasing

andronat commented:
It is already 1 commit. Do you see it differently?

andronat commented:
I can't see the building info from shippable, I am always getting "Build Not Found" error after 2-3 minutes of waiting.

bgrant0607 commented:
I see it as 1 commit now, thanks.

bgrant0607 commented:
The build is in progress: https://app.shippable.com/builds/55720e1f0d553e0c002dd044

We're investigating whether it's possible to relax the necessary privileges for seeing the logs on Shippable.

mikedanese commented:
I kicked it a couple minutes ago. It appeared to be a flake.

bgrant0607 commented:
A different flake. :-( Restarted tests again.

mikedanese commented:
4th try is the charm.
